casa/IO/MultiFile.cc

```diff
@@ -36,191 +37,752 @@
+    writeVector (cio, itsHdrCont[1].blockNrs);
+    itsNrContUsed[newContInx] = ncontNew;
+    cio.write (2, itsNrContUsed);
+    // Maybe write here actually used ncont0/1 !!!!!  Count those 8 bytes above
+    /// cio.write (2, itsncont);
+    // Write the continuation blocks.
+    const vector<Int64>& hdrContNrs = itsHdrCont[newContInx].blockNrs;
+    const uChar* remHdr = mio.getBuffer() + itsBlockSize;
+    todo = mio.length() - itsBlockSize;
+    if (itsTestMode != 0) {
+      cout << ' ' << newContInx << itsHdrCont[0].blockNrs << itsHdrCont[1].blockNrs
+           << hdrContNrs[0] << ' ' << ncont << ' ' << todo <<endl;
+    }
+    for (Int64 i=0; i<ncont; ++i) {
+      Int64 next = (i==ncont-1 ? Int64(0) : hdrContNrs[i+1]);
+      CanonicalConversion::fromLocal (iobuf, next);
+      memcpy (iobuf + sizeof(Int64), remHdr, std::min(todo,contBlkSize));
+      remHdr += contBlkSize;
+      todo -= contBlkSize;
+      itsIO->pwrite (itsBlockSize, hdrContNrs[i]*itsBlockSize, iobuf);
+    }
+    // Store first continuation blocknr in header.
+    uChar* buf = const_cast<uChar*>(mio.getBuffer());
+    CanonicalConversion::fromLocal (buf+8, hdrContNrs[0]);
+    // Swap continuation sets.
+    itsHdrContInx = newContInx;
+    // Store possibly changed nr of blocks.
+    CanonicalConversion::fromLocal (buf+48, itsNrBlock);
+  }
+
+  void MultiFile::writeVector (CanonicalIO& cio, const vector<Int64>& index)
+  {
+    // Write in the same way as AipsIO is doing.
+    Int64 sz = index.size();
+    cio.write (1, &sz);
+    if (sz > 0) {
+      cio.write (index.size(), &(index[0]));
+    }
+  }
+
+  void MultiFile::writeVector (CanonicalIO& cio, const vector<uInt>& index)
+  {
+    // Write in the same way as AipsIO is doing.
+    Int64 sz = index.size();
+    cio.write (1, &sz);
+    if (sz > 0) {
+      cio.write (index.size(), &(index[0]));
     }
-    todo -= itsBlockSize;
-    if (todo > 0) {
-      // The rest is written in another file. If the header info was written
-      // at the end of the file, it would be overwritten when extending with
-      // possible file corruption if the program or system crashes.
-      // By using a separate file, corruption chances are much lower.
-      // Even better would be using another name and doing a rename at the end.
-      int fd = RegularFileIO::openCreate (itsName + "_hdrext", ByteIO::New);
-      FiledesIO iohdr (fd, itsName + "_hdrext");
-      iohdr.write (todo, buf+itsBlockSize);
-      FiledesIO::close (fd);
+  }
+
+  void MultiFile::readVector (CanonicalIO& cio, vector<Int64>& index)
+  {
+    Int64 sz;
+    cio.read (1, &sz);
+    index.resize (sz);
+    if (sz > 0) {
+      cio.read (sz, &(index[0]));
+    }
+  }
+
+  void MultiFile::readVector (CanonicalIO& cio, vector<uInt>& index)
+  {
+    Int64 sz;
+    cio.read (1, &sz);
+    index.resize (sz);
+    if (sz > 0) {
+      cio.read (sz, &(index[0]));
     }
   }
 
   void MultiFile::readHeader (Bool always)
   {
```

Comment:
I think it would be better to have this function and two other functions
`readHeaderVersion1` and `readHeaderVersion2`. That would remove several ifs in this code and improve readability.

Reply:
Created them as suggested.